README.md
[![DOI](https://zenodo.org/badge/213679527.svg)](https://zenodo.org/badge/latestdoi/213679527)

# IEA-15-240-RWT v1.1 (Revised February 2022)
This repository contains the model data for the 15-MW offshore reference turbine developed within IEA Wind Task 37.

The documentation for the turbine is accessible here: https://www.nrel.gov/docs/fy20osti/75698.pdf
and the semisubmersible floating support structure is documented here: https://www.nrel.gov/docs/fy20osti/76773.pdf

Data in this repository includes:
* Documentation, including tabular data used in the figures from the technical report
* OpenFAST aeroelastic model inputs
* HAWC2 aeroelastic model inputs
* WISDEM optimization files
* Wind turbine ontology .yaml files
* CAD modeling of turbine where available

## Requirements

*OpenFAST*:
* Please check the release notes for OpenFAST version compatability.  OpenFAST can be compiled [from source here](https://github.com/OpenFAST/openfast.git) or precompiled Windows binaries are [available for download](https://github.com/OpenFAST/openfast/releases/latest/download/windows_openfast_binaries.zip). More information on installing and running OpenFAST is available in [OpenFAST documention](https://openfast.readthedocs.io/en/master/).
* NREL's Reference OpenSource Controller (ROSCO) is required.  This can be compiled [from source here](https://github.com/nrel/rosco) or precompiled binaries for all platforms are [available for download](https://github.com/NREL/ROSCO/releases/).  The version of the ROSCO controller can be found in the header of the [DISCON.IN](https://github.com/IEAWindTask37/IEA-15-240-RWT/blob/3a00f7f38a6373f6b026aef5878b671ca7af3605/OpenFAST/IEA-15-240-RWT-UMaineSemi/IEA-15-240-RWT-UMaineSemi_DISCON.IN#L2)

*HAWC2*:
* HAWC2 can be acquired from its [homepage](https://www.hawc2.dk/).  The DTU Basic Controller can be obtained from its [repository](https://gitlab.windenergy.dtu.dk/OpenLAC/BasicDTUController).

*WISDEM*:
 * WISDEM can be installed from its Github [repository](https://github.com/WISDEM/WISDEM).
 * See the [documentation](https://wisdem.readthedocs.io) for installation and usage guides.


## Design Updates

The IEA Wind Task 37 authors endeavor to keep the model input decks current with the latest releases and API changes.  Errors and other issues pointed out by the community are also addressed to the extent that available resources make that possible.  See the [Release Notes](blob/master/ReleaseNotes.md) for a detailed description of changes.

We also encourage the broader wind community to submit design updates by forking the repository and letting us know of your design customatization.  Community contributions that we are aware of include:
* [Bladed model](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)#is-bladed-supported) implemented by DNVGL, contact [William Collier](mailto:william.collier@dnv.com)
* [OrcaFlex model](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)#is-orcaflex-supported) implemented by Orcina, contact [Orcina](mailto:orcina@orcina.com)
* [Detailed rotor redesign](https://data.bris.ac.uk/data/dataset/3jrb4mejp9vfd2qb3s7dreymr1) from University of Bristol described in a TORQUE 2022 [paper](https://iopscience.iop.org/article/10.1088/1742-6596/2265/3/032029/pdf), contact [Peter Greaves](mailto:peter.greaves@ore.catapult.org.uk)
* [NuMAD model](https://github.com/UTDGriffithLab/UTD-IEA15MWBlade) developed at The University of Texas at Dallas (UTD) by [Alejandra S. Escalera Mendoza](mailto:ase180001@utdallas.edu) and [Prof D. Todd Griffith](mailto:tgriffith@utdallas.edu)
* [Jacket support structure](https://github.com/mmrocze2/IEA-15-240-RWT) The DEME Group created a 3-legged jacket for a 50m water depth, contact [Maciej Mroczek](mailto:Mroczek.Maciej@deme-group.com)

## Citations

For a list of academic papers that use or cite this turbine, please see [here (fixed-bottom)](https://scholar.google.com/scholar?cites=11739673662820715884&as_sdt=4005&sciodt=0,6&hl=en) and [here (floating)](https://scholar.google.com/scholar?cites=17665986740213390479&as_sdt=4005&sciodt=0,6&hl=en).

If you use this model in your research or publications, please cite the appropriate report as:

    @techreport{IEA15MW_ORWT,
    author = {Evan Gaertner and Jennifer Rinker and Latha Sethuraman and Frederik Zahle and Benjamin Anderson and Garrett Barter and Nikhar Abbas and Fanzhong Meng and Pietro Bortolotti and Witold Skrzypinski and George Scott and Roland Feil and Henrik Bredmose and Katherine Dykes and Matt Sheilds and Christopher Allen and Anthony Viselli},
    Howpublished = {NREL/TP-75698},
    institution = {International Energy Agency},
    title = {Definition of the {IEA} 15-Megawatt Offshore Reference Wind Turbine},
    URL = {https://www.nrel.gov/docs/fy20osti/75698.pdf},
    Year = {2020}
    }

    @techreport{IEA15MW_ORWT_Floating,
    author = {Christopher Allen and Anthony Viselli and Habib Dagher and Andrew Goupee and Evan Gaertner and Nikhar Abbas and Matthew Hall and Garrett Barter},
    Howpublished = {NREL/TP-76773},
    institution = {International Energy Agency},
    title = {Definition of the {UMaine} {VolturnUS-S} Reference Platform Developed for the {IEA Wind} 15-Megawatt Offshore Reference Wind Turbine}},
    URL = {https://www.nrel.gov/docs/fy20osti/76773.pdf},
    Year = {2020}
    }

## Questions

Before reaching out to NREL or DTU authors with questions on the model or reports, please see our frequently asked questions (FAQ) on our [Github Wiki](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)) and current or prior [Issues](https://github.com/IEAWindTask37/IEA-15-240-RWT/issues).

If neither the FAQ or Issues address your need, please create a new Issue on this repository so that the dialogue is archived for others that might have similar questions. You can also reach out to the authors directly if that is your preference.  The technical report lists the contributions of individual authors if you have a specific question.  Otherwise, you can contact Garrett Barter (garrett.barter@nrel.gov).


ReleaseNotes.md
# IEA Wind 15-MW Release Notes

## v 1.1.6

 * Addition of HAWC2 monopile model
 
 * Alignment of tower and monopile discretizations and properties across YAML, HAWC2, and OpenFAST files
 
 * Addition of SolidWorks CAD model for UMaine VolturnUS-S semisubmersible 

## v 1.1.5

 * Update OpenFAST input files for v3.5.1 release API changes
 
 * Floating tower damping reverted to 1.0% to resolve numerical instabilities. There is a good deal of uncertainty about the true value here as it is normally determined experimentally.  Comments to this effect have been added to the [Wiki entry](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)#what-are-the-values-of-structural-damping).

 * Update the second-order potential-flow wave excitation of the VolturnUS-S platform.  [PR #137](https://github.com/IEAWindTask37/IEA-15-240-RWT/pull/137).

## v 1.1.4

 * Update OpenFAST input files for v3.4.1 release API changes
 
 * Update ROSCO libraries and inputs to v2.7
 
 * Blade elastic properties consistently originate from SONATA/ANBA4

 * Cleanup of stale files and README comments
 
## v 1.1.3

 * Update OpenFAST input files for v3.3 release API changes
 
 * Update ROSCO to v2.6
 
 * Provide both WAMIT input and output files, with a bug fix regarding platform orientation
 
 * Adjust OpenFAST ElastoDyn NacYIner and PtfmYIner according to monopile simulation best practices 
 
 * Additional references to community design contributions and permutations in the README
 
 * Remove deprecated BModes files
 
 * Fix minor errors in the tabulated documentation data

## v 1.1.2

 * Update ROSCO to v2.5

## v 1.1.1

 * Fix OpenFAST floating tower mode shape coefficients to avoid instability
 
 * Fix incorrect OpenFAST VolturnUS-S PtfmCMzt value to WISDEM-generated value and closer to original value

## v 1.1

This update to the IEA Wind 15-MW Reference Wind Turbine attempts to address some of the modeling issues that have been pointed out by users.  Larger redesign requests, such as a smaller diameter monopile, or higher specific power for the floating turbine, are not included as that would require a more significant allocation of resources and will likely be left to the broader community to generate design alternatives.  Please see the [Wiki FAQ](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)) for some extended answers to those topics.

**Major Changes**

 * Many of the modeling tools have undergone changes and updates, so the files here have been updated to keep pace with API changes.

   - WISDEM has undergone perhaps the most extensive overhaul.  It now uses the YAML-based ontology files from the [WindIO](https://github.com/IEAWindTask37/windIO) project as the chief input, and is also more tightly integrated with OpenFAST and ROSCO.  More detailed changes described below.

   - OpenFAST has seen a couple of releases since the 15-MW RWT input files were created.  Additionally, the original input deck also had a number of errors and inconsistencies.  The integrations with WISDEM and, by extension, the WindIO ontology files have helped to rectify these problems.  Finally, a set of input files that works with OLAF, the OpenFAST vortex particle aerodynamics code has been included.

   - ROSCO, which provides the controller, was a new tool when the 15-MW RWT was first released and has since seen numerous bug fixes and new features.  More details described below and at the [ROSCO repository](https://github.com/nrel/rosco).

 * The WindIO ontology YAML-files have been expanded to include descriptions of the entire turbine.  This follows the progress made in the parallel IEA Task 37 [WindIO](https://github.com/IEAWindTask37/windIO) project.  Note that the YAML descriptions for some of these components will likely still evolve, especially the nacelle components, as settling on a communal parameterization takes time and numerous iterations.

 * Draft version of the ontology for the UMaine Volturn-S semisubmersible design variant is now included in the `WT_Ontology` directory.  This should be considered a work in-progress as there is still extensive work to do to verify how well the description matches the original intent of the design report.

 * There was an inconsistency in the stated hub diameter in the original report, YAML-definition, and WISDEM files (7.94m) versus the hub diameter in the OpenFAST and HAWC2 files (3.0m).  This was due to a late design change to better accommodate the blade root chord/diameter size that did not fully propagate to all input files.  With the hub diameter at 7.94m and the blade unchanged at 117.15m in length (arc length measurement that includes the prebend), the precise rotor diameter is 242.24m.  This is called out in [Issue #51](https://github.com/IEAWindTask37/IEA-15-240-RWT/issues/51) and the [Wiki FAQ](https://github.com/IEAWindTask37/IEA-15-240-RWT/wiki/Frequently-Asked-Questions-(FAQ)).

 * The nacelle component mass properties and dimensions were approximate in the report and then inconsistently captured between the various model input decks.  Furthermore Table 5-1 in the report had values that could do be replicated and did not add up (called out in [Issue #30](https://github.com/IEAWindTask37/IEA-15-240-RWT/issues/30).  With the new capabilities in WISDEM's DrivetrainSE model, this has been rectified in the nacelle mass properties tab in the IEA-15-240-RWT_tabular.xlsx file in the [Documentation](https://github.com/IEAWindTask37/IEA-15-240-RWT/tree/master/Documentation) folder.


WISDEM/run_model.py
import os
from collections import OrderedDict
from wisdem import run_wisdem
import wisdem.postprocessing.compare_designs as compare_designs
import wisdem.postprocessing.wisdem_get as getter
import wisdem.commonse.utilities as util
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from generateTables import RWT_Tabular

# File management
thisdir = os.path.dirname(os.path.realpath(__file__))
ontology_dir = os.path.join(os.path.dirname(thisdir), "WT_Ontology")
fname_modeling_options = os.path.join(thisdir, "modeling_options.yaml")
fname_analysis_options = os.path.join(thisdir, "analysis_options.yaml")
folder_output = os.path.join(thisdir, "outputs")

def run_15mw(fname_wt_input):
    float_flag = fname_wt_input.find('Volturn') >= 0
    
    # Run WISDEM
    prob, modeling_options, analysis_options = run_wisdem(fname_wt_input, fname_modeling_options, fname_analysis_options)

    # Produce standard plots
    compare_designs.run([prob], ['IEA Wind 15-MW'], modeling_options, analysis_options)

    # Tabular output: Blade Shape
    blade_shape  = np.c_[prob.get_val('blade.outer_shape_bem.s'),
                         prob.get_val('blade.outer_shape_bem.ref_axis','m')[:,2],
                         prob.get_val('blade.outer_shape_bem.chord','m'),
                         prob.get_val('blade.outer_shape_bem.twist', 'deg'),
                         prob.get_val('blade.interp_airfoils.r_thick_interp')*100,
                         prob.get_val('blade.outer_shape_bem.pitch_axis')*100,
                         prob.get_val('blade.outer_shape_bem.ref_axis','m')[:,0],
                         prob.get_val('blade.outer_shape_bem.ref_axis','m')[:,1],
                         ]
    blade_shape_col = ['Blade Span','Rotor Coordinate [m]',
                       'Chord [m]', 'Twist [deg]',
                       'Relative Thickness [%]', 'Pitch Axis Chord Location [%]',
                       'Prebend [m]', 'Sweep [m]']
    bladeDF = pd.DataFrame(data=blade_shape, columns=blade_shape_col)

    # Tabular output: Blade Stiffness
    blade_stiff  = np.c_[prob.get_val('rotorse.r','m'),
                         prob.get_val('rotorse.A','m**2'),
                         prob.get_val('rotorse.EA','N'),
                         prob.get_val('rotorse.EIxx','N*m**2'),
                         prob.get_val('rotorse.EIyy','N*m**2'),
                         prob.get_val('rotorse.EIxy','N*m**2'),
                         prob.get_val('rotorse.GJ','N*m**2'),
                         prob.get_val('rotorse.rhoA','kg/m'),
                         prob.get_val('rotorse.rhoJ','kg*m'),
                         prob.get_val('rotorse.x_ec','mm'),
                         prob.get_val('rotorse.y_ec','mm'),
                         prob.get_val('rotorse.re.x_tc','mm'),
                         prob.get_val('rotorse.re.y_tc','mm'),
                         prob.get_val('rotorse.re.x_sc','mm'),
                         prob.get_val('rotorse.re.y_sc','mm'),
                         prob.get_val('rotorse.re.x_cg','mm'),
                         prob.get_val('rotorse.re.y_cg','mm'),
                         prob.get_val('rotorse.re.precomp.flap_iner','kg/m'),
                         prob.get_val('rotorse.re.precomp.edge_iner','kg/m')]
    blade_stiff_col = ['Blade Span [m]',
                       'Cross-sectional area [m^2]',
                       'Axial stiffness [N]',
                       'Edgewise stiffness [Nm^2]',
                       'Flapwise stiffness [Nm^2]',
                       'Flap-edge coupled stiffness [Nm^2]',
                       'Torsional stiffness [Nm^2]',
                       'Mass density [kg/m]',
                       'Polar moment of inertia density [kg*m]',
                       'X-distance to elastic center [mm]',
                       'Y-distance to elastic center [mm]',
                       'X-distance to tension center [mm]',
                       'Y-distance to tension center [mm]',
                       'X-distance to shear center [mm]',
                       'Y-distance to shear center [mm]',
                       'X-distance to mass center [mm]',
                       'Y-distance to mass center [mm]',
                       'Section flap inertia [kg/m]',
                       'Section edge inertia [kg/m]',
                       ]
    bladeStiffDF = pd.DataFrame(data=blade_stiff, columns=blade_stiff_col)

    # Blade internal laminate layer details
    layerDF = []
    l_s = prob.get_val("blade.internal_structure_2d_fem.s")
    lthick = prob.get_val("blade.internal_structure_2d_fem.layer_thickness", 'm')
    lorient = prob.get_val("blade.internal_structure_2d_fem.layer_orientation", 'deg')
    lstart = prob.get_val("blade.internal_structure_2d_fem.layer_start_nd")
    lend = prob.get_val("blade.internal_structure_2d_fem.layer_end_nd")
    nlay = lthick.shape[0]
    layer_cols = ['Span','Thickness [m]','Fiber angle [deg]','Layer Start','Layer End']
    for k in range(nlay):
        ilay = np.c_[l_s, lthick[k,:], lorient[k,:], lstart[k,:], lend[k,:]]
        layerDF.append( pd.DataFrame(data=ilay, columns=layer_cols) )
    
    # Tabular output: Rotor Performance
    rotor_perf = np.c_[prob.get_val("rotorse.rp.powercurve.V",'m/s'),
                       prob.get_val("rotorse.rp.powercurve.pitch",'deg'),
                       prob.get_val("rotorse.rp.powercurve.P",'MW'),
                       prob.get_val("rotorse.rp.powercurve.Cp"),
                       prob.get_val("rotorse.rp.powercurve.Cp_aero"),
                       prob.get_val("rotorse.rp.powercurve.Omega",'rpm'),
                       prob.get_val("rotorse.rp.powercurve.Omega",'rad/s')*0.5*prob["configuration.rotor_diameter_user"],
                       prob.get_val("rotorse.rp.powercurve.T",'MN'),
                       prob.get_val("rotorse.rp.powercurve.Ct_aero"),
                       prob.get_val("rotorse.rp.powercurve.Q",'MN*m'),
                       prob.get_val("rotorse.rp.powercurve.Cq_aero"),
                       prob.get_val("rotorse.rp.powercurve.M",'MN*m'),
                       prob.get_val("rotorse.rp.powercurve.Cm_aero"),
                       ]
    rotor_perf_col = ['Wind [m/s]','Pitch [deg]',
                      'Power [MW]','Power Coefficient [-]','Aero Power Coefficient [-]',
                      'Rotor Speed [rpm]','Tip Speed [m/s]',
                      'Thrust [MN]','Thrust Coefficient [-]',
                      'Torque [MNm]','Torque Coefficient [-]',
                      'Blade Moment [MNm]','Blade Moment Coefficient [-]',
                      ]
    perfDF = pd.DataFrame(data=rotor_perf, columns=rotor_perf_col)

    # Nacelle mass properties tabular
    # Columns are ['Mass', 'CoM_x', 'CoM_y', 'CoM_z',
    #              'MoI_cm_xx', 'MoI_cm_yy', 'MoI_cm_zz', 'MoI_cm_xy', 'MoI_cm_xz', 'MoI_cm_yz',
    #              'MoI_TT_xx', 'MoI_TT_yy', 'MoI_TT_zz', 'MoI_TT_xy', 'MoI_TT_xz', 'MoI_TT_yz']
    nacDF = prob.model.wt.drivese.nac._mass_table
    hub_cm = float(prob["drivese.hub_system_cm"])
    L_drive = float(prob["drivese.L_drive"])
    tilt = float(prob.get_val('nacelle.uptilt', 'rad'))
    shaft0 = prob["drivese.shaft_start"]
    Cup = -1.0
    hub_cm = R = shaft0 + (L_drive + hub_cm) * np.array([Cup * np.cos(tilt), 0.0, np.sin(tilt)])
    hub_mass = prob['drivese.hub_system_mass']
    hub_I = prob["drivese.hub_system_I"]
    hub_I_TT = util.rotateI(hub_I, -Cup * tilt, axis="y")
    hub_I_TT = util.unassembleI( util.assembleI(hub_I_TT) +
                                 hub_mass * (np.dot(R, R) * np.eye(3) - np.outer(R, R)) )
    blades_mass = prob['drivese.blades_mass']
    blades_I = prob["drivese.blades_I"]
    blades_I_TT = util.rotateI(blades_I, -Cup * tilt, axis="y")
    blades_I_TT = util.unassembleI( util.assembleI(blades_I_TT) +
                                    blades_mass * (np.dot(R, R) * np.eye(3) - np.outer(R, R)) )
    rna_mass = prob['drivese.rna_mass']
    rna_cm = R = prob['drivese.rna_cm']
    rna_I_TT = prob['drivese.rna_I_TT']
    rna_I = util.unassembleI( util.assembleI(rna_I_TT) +
                                    rna_mass * (np.dot(R, R) * np.eye(3) + np.outer(R, R)) )
    nacDF.loc['Blades'] = np.r_[blades_mass, hub_cm, blades_I, blades_I_TT].tolist()
    nacDF.loc['Hub_System'] = np.r_[hub_mass, hub_cm, hub_I, hub_I_TT].tolist()
    nacDF.loc['RNA'] = np.r_[rna_mass, rna_cm, rna_I, rna_I_TT].tolist()

    # Tabular output: Tower
    water_depth = prob['env.water_depth']
    h_trans = getter.get_transition_height(prob)
    htow = getter.get_zpts(prob) #np.cumsum(np.r_[0.0, prob['towerse.tower_section_height']]) + prob['towerse.z_start']
    t = getter.get_tower_thickness(prob)
    towdata = np.c_[htow,
                    getter.get_tower_diameter(prob),
                    np.r_[t[0], t]]
    rowadd = []
    for k in range(towdata.shape[0]):
        if k==0: continue
        if k+1 < towdata.shape[0]:
            rowadd.append([towdata[k,0]+1e-3, towdata[k,1], towdata[k+1,2]])
    towdata = np.vstack((towdata, rowadd))
    towdata[:,-1] *= 1e3
    towdata = np.round( towdata[towdata[:,0].argsort(),], 3)
    colstr = ['Height [m]','OD [m]', 'Thickness [mm]']
    towDF = pd.DataFrame(data=towdata, columns=colstr)
    mycomments = ['']*towdata.shape[0]
    if not float_flag:
        #breakpoint()
        mycomments[0] = 'Monopile start'
        mycomments[np.where(towdata[:,0] == -water_depth)[0][0]] = 'Mud line'
        mycomments[np.where(towdata[:,0] == 0.0)[0][0]] = 'Water line'
    idx_tow = np.where(towdata[:,0] == h_trans)[0][0]
    mycomments[idx_tow] = 'Tower start'
    mycomments[-1] = 'Tower top'
    towDF['Location'] = mycomments
    towDF = towDF[['Location']+colstr]
    A = 0.25*np.pi*(towDF['OD [m]']**2 - (towDF['OD [m]']-2*1e-3*towDF['Thickness [mm]'])**2)
    I = (1/64.)*np.pi*(towDF['OD [m]']**4 - (towDF['OD [m]']-2*1e-3*towDF['Thickness [mm]'])**4)
    outfitting = np.zeros( len(A) )
    if not float_flag:
        outfitting[:idx_tow] = prob['fixedse.outfitting_factor_in']
    outfitting[idx_tow:] = prob['towerse.outfitting_factor_in']
    towDF['Mass Density [kg/m]'] = outfitting * getter.get_tower_rho(prob)[0] * A
    towDF['Fore-aft inertia [kg.m]'] = towDF['Side-side inertia [kg.m]'] = towDF['Mass Density [kg/m]'] * I/A
    towDF['Fore-aft stiffness [N.m^2]'] = towDF['Side-side stiffness [N.m^2]'] = getter.get_tower_E(prob)[0] * I
    towDF['Torsional stiffness [N.m^2]'] = getter.get_tower_G(prob)[0] * 2*I
    towDF['Axial stiffness [N]'] = getter.get_tower_E(prob)[0] * A
    #with open('tow.tbl','w') as f:
    #    towDF.to_latex(f, index=False)

    # Frequency plot
    fn_tower = getter.get_tower_freqs(prob)
    f        = np.linspace(0., 0.5, num=1000)[1:]
    omega    = f*(2.*np.pi)

    Omega_rot_min = prob.get_val('control.minOmega', 'rpm')/60.
    Omega_rot_max = prob.get_val('control.maxOmega', 'rpm')/60.

    f_1P = np.r_[0, Omega_rot_min, Omega_rot_min, Omega_rot_max, Omega_rot_max, f[-1]]
    f_3P = np.r_[0, 3.*Omega_rot_min, 3.*Omega_rot_min, 3.*Omega_rot_max, 3.*Omega_rot_max, f[-1]]
    NP_y = np.r_[0., 0., 1., 1., 0., 0.]

    # f_1P_marg_low  = Omega_rot_min/1.1
    # f_1P_marg_high = Omega_rot_min*1.1
    # f_3P_marg_low  = Omega_rot_min*3./1.1
    # f_3P_marg_high = Omega_rot_min*3.*1.1

    f_tower = [fn_tower[0], fn_tower[0]]
    y_tower = [0., 1.]

    # Kaimal
    Sigma1 = 42
    V_hub  = 10.
    L_k = 8.1*Sigma1
    Sk = np.zeros_like(f)
    for i, fi in enumerate(f):
        Sk[i] = 4*fi*L_k/V_hub/(1+6*fi*L_k/V_hub)**(5./3.)/fi
    Sk /= max(Sk)

    # Pierson-Moskowitz
    F = 500000
    g = 9.81
    U10 = 10.

    U19     = U10*1.17
    alpha   = 8.1e-3
    Beta    = 0.74
    omega_0 = g/U19
    omega_p = 0.877*g/U19

    S = np.zeros_like(f)
    for i, omega_i in enumerate(omega):
        S[i] = alpha*g**2./omega_i**5.*np.exp(-Beta*(omega_0/omega_i)**4.)
    S /= max(S)

    fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(6., 2.5))
    fig.subplots_adjust(bottom=0.36, top=0.975,  left=0.12, right=0.975)#, right=0.8, hspace=0.3, wspace=0.3)

    ax.fill(f_1P, NP_y, color=[0.5,0.5,0.5], label='1P')
    ax.fill(f_3P, NP_y, color=[0.75,0.75,0.75], label='3P')
    ax.plot(f, Sk, color='r', label='Wind,\nKaimal Spect.', linewidth=1.5)
    ax.plot(f, S,  color='b', label="Waves,\nJONSWAP Spect.", linewidth=1.5)
    ax.plot(f_tower, y_tower, color='k', label="Tower,\n1st Nat. Freq.", linewidth=1.5)
    ax.grid(color=[0.8,0.8,0.8], linestyle='--')
    ax.set_ylabel('Normalized PSD')
    ax.set_xlabel('Frequency [Hz]')
    ax.set_xlim((0.,max(f)))

    ax.set_axisbelow(True)

    handles, labels = ax.get_legend_handles_labels()
    fig.legend(handles, labels, loc='upper center', bbox_to_anchor=(0.5, 0.2), ncol=5, prop={'size':9})
    fig_name = 'tower_fn'
    fig.savefig(folder_output + os.sep + fig_name+'.pdf', pad_inches=0.1, bbox_inches='tight')
    fig.savefig(folder_output + os.sep + fig_name+'.png', pad_inches=0.1, bbox_inches='tight')

    # Overview dictionary
    overview = OrderedDict()

    overview['Power rating [MW]'] = prob.get_val('configuration.rated_power','MW')
    overview['Turbine class'] = 'IEC Class '+ prob['configuration.ws_class']+prob['configuration.turb_class']
    overview['Rotor diameter [m]'] = prob["configuration.rotor_diameter_user"]
    overview['Specific rating [W/m^2]'] = prob.get_val('configuration.rated_power','W')/np.pi/(0.5*prob["configuration.rotor_diameter_user"])**2.0
    overview['Rotor orientation'] = 'Upwind' if prob['configuration.upwind'] else 'Downwind'
    overview['Number of blades'] = prob['configuration.n_blades']
    overview['Control'] = 'Variable speed, Collective pitch'
    overview['Cut-in wind speed [m/s]'] = prob['control.V_in']
    overview['Rated wind speed [m/s]'] = prob["rotorse.rp.powercurve.rated_V"]
    overview['Cut-out wind speed [m/s]'] = prob['control.V_out']
    overview['Airfoil series'] = 'FFA-W3'
    overview['Hub height [m]'] = prob['configuration.hub_height_user']
    overview['Hub diameter [m]'] = prob['hub.diameter']
    overview['Hub Overhang [m]'] = -prob['nacelle.overhang']
    overview['Drive train'] = 'Low speed, Direct drive'
    overview['Design tip speed ratio'] = prob['control.rated_TSR']
    overview['Minimum rotor speed [rpm]'] = prob.get_val('control.minOmega','rpm')
    overview['Maximum rotor speed [rpm]'] = prob.get_val('control.maxOmega','rpm')
    overview['Maximum tip speed [m/s]'] = prob['control.max_TS']
    overview['Shaft tilt angle [deg]'] = prob.get_val('nacelle.uptilt','deg')
    overview['Rotor cone angle [deg]'] = prob.get_val('hub.cone','deg')
    overview['Tower top to hub flange height [m]'] = prob['nacelle.distance_tt_hub']
    overview['Generator rated efficiency [%]'] = prob['rotorse.rp.powercurve.rated_efficiency']
    overview['Blade pre-bend [m]'] = prob['blade.outer_shape_bem.ref_axis'][-1,0]
    overview['Blade mass [t]'] = 1e-3*prob['rotorse.re.precomp.blade_mass']
    overview['Hub mass [t]'] = 1e-3*prob['drivese.hub_mass']
    overview['Generator mass [t]'] = 1e-3*prob['drivese.generator_mass']
    overview['Nacelle mass [t]'] = 1e-3*prob['drivese.nacelle_mass']
    overview['RNA mass [t]'] = 1e-3*prob['drivese.rna_mass']
    overview['Tower mass [t]'] = 1e-3*getter.get_tower_mass(prob)
    overview['Tower base diameter [m]'] = towdata[np.where(towdata[:,0] == h_trans)[0][0],1]
    overview['Transition piece height [m]'] = h_trans
    if not float_flag:
        overview['Monopile embedment depth [m]'] = prob['fixedse.suctionpile_depth']
        overview['Monopile base diameter [m]'] = prob['fixedse.monopile_outer_diameter'][0]
        overview['Monopile mass [t]'] = 1e-3*prob['fixedse.monopile_mass']
    else:
        overview['Volturn-S hull mass [t]'] = 1e-3*prob['floatingse.platform_hull_mass']
        overview['Volturn-S fixed ballast mass [t]'] = 1e-3*prob['floatingse.platform_ballast_mass'][0]
        overview['Volturn-S fluid ballast mass [t]'] = 1e-3*prob['floatingse.variable_ballast_mass'][0]
        overview['Volturn-S displacement [m^3]'] = prob['floatingse.platform_displacement']
        overview['Volturn-S freeboard [m]'] = 15.0
        overview['Volturn-S draft [m]'] = 20.0

    # Write all tabular data to xlsx
    myobj = RWT_Tabular(fname_wt_input, towDF=towDF, rotDF=perfDF,
                        nacDF=nacDF, layerDF=layerDF, overview=overview)
    myobj.write_all()


if __name__ == '__main__':
    run_15mw( os.path.join(ontology_dir, "IEA-15-240-RWT.yaml") )
    run_15mw( os.path.join(ontology_dir, "IEA-15-240-RWT_VolturnUS-S.yaml") )
    
